drivers/dma/dma_cavs.c

```diff
@@ -233,10 +233,18 @@ static int dw_dma_config(struct device *dev, u32_t channel,
 		cnt--;
 	} while (cfg_blocks && cnt);
```

Comment:
I am assuming "cnt" would be initialized appropriately in the App layer in case of a cyclic buffer. i.e. if the buffer contains 4 blocks, "cnt" should be 4. Also, cfg_blocks should be set up in a circular manner in the App layer.

Reply:
Yes, correct. `cnt` will be initialized according to the total number of blocks by the application. This is being done even without this patch.
Applications requesting circular list operation will indicate this mode via `cfg_blocks` that has a valid next pointer in the last block. However, it is _technically_ not necessary for the last block to point to the first block, the application can simply set the next point to be non-NULL. We can expect applications to link the block to a valid block in the list rather than to set a random non-NULL value though.

Reply:
I am just trying to emphasize the importance of cnt here. If the number of blocks is 4 and if the cnt is given as 6 with the last block in cfg_blocks pointing back to the first block, the loop will run for 6 times. And then when it comes out of the loop, cfg_blocks is again made to point to the first block potentially messing up the sequence (first 2 blocks will be DMAed twice). I just wanted you to be aware of this scenario and if possible prevent it in the App layer.

Reply:
Agreed. To be fair, the consistency between config_block pointers and the count was needed even before the patch. Consider a case before my patch where the DMA's application provided 4 instances of config_blocks in a circular list and provided the count as 6. The code would not have detected this and created 6 LLIs.
Since the DMA will be used by other peripheral drivers, we need ensure that these drivers are implemented with this constraint in mind. Currently the only user of this DMA is the I2S driver and unfortunately, the current I2S APIs do not allow for a circular list in a clean way, so I have not modified the I2S driver to use circular lists. In case of the PDM driver, the constraint is being taken care of.